### ISSUE TITLE
Date range capability in backtester

### ISSUE THREAD

**Issue by ryankennedyio** [mhallsmoore/qstrader#15]
Rather simple, but will be quite helpful to only run a backtest through a period we're interested in. Will also help when doing simple in-sample/out-of-sample comparisons.

I believe this should be a function of the backtester rather than price handler, as it's a function specific to backtesting, however this might throw up some performance issues if someone is doing second-resolution backtesting on a large stock universe ?

**Comment by ryankennedyio** [comment]
Working version here;

https://github.com/ryankennedyio/qstrader/commit/4e3c76f6761c34deef62ef9ca6cc2c05433c6d44

As suspected.. way too slow, even on a stock universe of ~300 with daily data going back to 2000

**Comment by ryankennedyio** [comment]
After setting this up the 'dumb' way, I've realised it needs to be a function of the price handler for speed's sake. Even then it's kinda like watching grass grow until the date-window shows up.


@mhallsmoore I've also noted the need to move much of the backtest setup outside of the `Backtest.init()` function for the sake of easy parameter setup for each component of the system.
This way, I have a `working.py` file in, say `qstrader/workdir`. Here we pretty much do what `Backtest.init()` does, but it also gives us the opportunity to say, do `price_handler.set_date_window()` before we start running the actual backtest.

Basic idea is here. I haven't yet set up the example backtests to work with this, only my own.
https://github.com/ryankennedyio/qstrader/commit/8ece3ddb675168971ace866b2b82291a3da621b5

What do you think is appropriate?



Either way I think this will be too slow with the current way of feeding in ticks from CSV files (?) -- are you planning on including a different historical DB, or leaving it as an exercise for the reader? Only reason I ask is that if one is planned, I'll help setting that up as my own starting point.

**Comment by mhallsmoore** [comment]
We'll definitely need to use a different historical DB, as a separate price handler, pretty soon. Once we start date-ranging high volume time series data over a large stock universe, it will be necessary to use something akin to Cassandra or HDF5. I've not had experience with the former, but I have some with the latter.

I believe QSTrader should (eventually) be relatively agnostic to the back-end, so perhaps we could start with HDF5 and then maybe a Cassandra price handler later?

I believe, at least for tick data, a common way to store multi-day stock universes is one-file-per-stock-per-day of tick data. There was a great discussion kicking around somewhere about how to store this, I'll try dig it out.

Ultimately we should try and keep the "example.py" files as short as possible (although some complexity is obviously unavoidable) and keep boilerplate in the Backtest or PriceHandler libs if we can.

**Comment by ryankennedyio** [comment]
Cool, on that note I'll go about using HDF5 -- I've toyed with Cassandra and liked it, but HDF5 seems well regarded within the scientific/python community.

Here was one example I ran across that may be useful:
https://dcaisen.wordpress.com/2012/09/11/storing-and-analyzing-equity-tick-data-trades-and-quotes-using-hdf5-h5py/

Either way, whatever DB the person uses is abstracted behind a price handler, so I think it's got a very backend-agnostic design. Even something more exotic like kdb would be a breeze, just write the price handler to suit it.


I've set Yahoo price handler up to deal with time-window constraints now quite efficiently by splicing the tick_stream Dataframe according to date indexes before we start streaming them in. 

Example SP500 strat with the `Backtest.init()` moved into the working/example files:

https://github.com/ryankennedyio/qstrader/blob/backtest-date-constraints/examples/test_sp500tr_buy_and_hold_backtest.py

I'll run all the tests again and make sure before submitting a PR, but that's working pretty nicely for what I'm looking at.

I  totally agree that boilerplate should be avoided, but I was struggling to see any alternatives for setting the individual component parameters. Maybe there's a way to do this that I'm not yet seeing in Python?
